kitchen/src/kitchen/demos.clj

```diff
@@ -0,0 +1,144 @@
+           (java.net URLEncoder)))
+
+(defn- request->cid [request]
+  (get-in request [:headers "x-cid"]))
+
+(defn- request->cid-string [request]
+  (str "[CID=" (request->cid request) "]"))
+
+(defn- printlog
+  ([request & messages]
+   (log/info (request->cid-string request) (str/join " " messages))))
+
+(defn- search->image-titles
+  [search-query]
+  (let [extract-titles (fn extract-titles [page-entries]
+                         (mapcat (fn [page-entry] (map #(get % "title") (get page-entry "images"))) page-entries))
+        search-url (str "https://en.wikipedia.org/w/api.php?action=query&prop=images&format=json&titles=" (URLEncoder/encode search-query))
+        initial-result (-> search-url slurp json/read-str)]
+    (-> initial-result
+        (get-in ["query" "pages"])
+        vals
+        extract-titles)))
+
+(defn- extract-entries
+  [page-entries]
+  (-> (map (fn [page-entry]
+             {"title" (get page-entry "title")
+              "url" (-> page-entry
+                        (get "imageinfo")
+                        first
+                        (get "url"))})
+           page-entries)
+      vec
+      first))
+
+(defn- image-title->image-entry
+  [image-title]
+  (-> (str "https://en.wikipedia.org/w/api.php?action=query&prop=imageinfo&iiprop=url&format=json&titles=" (URLEncoder/encode image-title))
+      slurp
+      json/read-str
+      (get-in ["query" "pages"])
+      vals
+      extract-entries))
+
+(defn- image-search-handler
+  ""
+  [{:keys [query-params] :as request}]
+  (printlog request "query params:" query-params)
+  (try
+    (let [{:strs [search]} query-params
+          search-url (str "https://en.wikipedia.org/w/api.php?action=query&prop=images&format=json&titles=" search)
+          image-titles (search->image-titles search)
+          _ (printlog request "image titles:" (vec image-titles))
+          image-entries (map image-title->image-entry image-titles)]
+      (printlog request "image entries:" (vec image-entries))
+      (utils/map->json-response {:result image-entries}))
+    (catch Exception e
+      (log/error e "encountered exception while performing image search")
+      (utils/exception->json-response e))))
+
+(defn- directory-location
+  [dir-name]
+  (let [base-dir (or (System/getenv "MESOS_DIRECTORY") (System/getProperty "user.dir"))]
+    (str base-dir File/separator dir-name)))
+
+(defn- copy-uri-to-file
+  [uri file]
+  (with-open [in (io/input-stream uri)
+              out (io/output-stream file)]
+    (io/copy in out)))
+
+(let [image-tagging-lock (Object.)]
+  (defn- retrieve-image-tags
+    [classifier-file image-location predictions]
+    (let [result-atom (atom nil)]
+      (locking image-tagging-lock
+        (reset! result-atom (sh/sh "python3"
+                                   classifier-file
+                                   (str "--image_file=" image-location)
+                                   (str "--num_top_predictions=" predictions))))
+      @result-atom)))
```

Comment:
I don't understand why we need the atom

Reply:
We need to return the result of the `sh` function that has the exit code, stdout and stderr.